File: starters/typescript/workflow-runner/src/runner/workflow-runner.ts
/**
 * Workflow Runner - Job Execution Engine
 *
 * Provides utilities for executing jobs (workflow steps) in a controlled manner.
 * This module is the core of the workflow-runner architecture.
 *
 * ## Responsibilities
 * - Execute jobs in defined order
 * - Control success/failure flow
 * - Handle process exit codes
 *
 * ## NOT Responsible For
 * - Parallel execution
 * - Retry logic
 * - Skip logic
 * - Interactive prompts
 * - Job scheduling (cron handles this externally)
 *
 * ## Design Philosophy
 * Jobs are workflow steps. They are executed sequentially by this runner.
 * The runner does not make decisions about which jobs to run - that is
 * determined by the workflow definition (index.ts or external orchestrator).
 *
 * @module lib/runner
 */

import { fileURLToPath } from "url";
import { argv } from "process";
import type { JobResult, JobFunction } from "./types.js";
import logger from "../infra/logger/pino.js";
import { EXIT_CODES } from "../shared/constants.js";

/**
 * Check if current module is the main entry point (ESM version)
 * Equivalent to `require.main === module` in CommonJS
 *
 * This is useful for files that can be both imported as a module
 * and executed directly as a CLI script.
 *
 * @param importMetaUrl - Pass `import.meta.url` from the calling module
 * @returns true if the module is being run directly, false if imported
 *
 * @example
 * ```typescript
 * if (isMainModule(import.meta.url)) {
 *   console.log("Running as CLI");
 * } else {
 *   console.log("Imported as module");
 * }
 * ```
 */
export function isMainModule(importMetaUrl: string): boolean {
  const modulePath = fileURLToPath(importMetaUrl);
  const entryPoint = argv[1];
  return modulePath === entryPoint;
}

/**
 * Execute a job (workflow step) with proper exit handling
 *
 * This is the primary entry point for running a job as a workflow step.
 * Only calls process.exit() when running as main module (not when imported).
 *
 * ## Workflow Context
 * Jobs are workflow steps. When executed directly (via npm script or node),
 * this function handles the process lifecycle. When imported (Lambda, tests),
 * it returns without calling process.exit().
 *
 * ## Exit Codes
 * - 0: Job succeeded (result.success === true)
 * - 1: Job failed (result.success === false or uncaught error)
 *
 * @param job - The job function (workflow step) to execute
 * @param importMetaUrl - Pass `import.meta.url` from the calling module
 *
 * @example
 * ```typescript
 * // At the end of your job file:
 * export const main = myJob.handler;
 * runIfMain(main, import.meta.url);
 * ```
 */
export function runIfMain(job: JobFunction, importMetaUrl: string): void {
  if (!isMainModule(importMetaUrl)) {
    return;
  }

  job()
    .then((result: JobResult) => {
      process.exit(result.success ? EXIT_CODES.SUCCESS : EXIT_CODES.FAILURE);
    })
    .catch((error: unknown) => {
      logger.fatal({ error }, "Unexpected error");
      process.exit(EXIT_CODES.FAILURE);
    });
}

File: starters/rust-batch/src/runner/workflow_runner.rs
//! Workflow Runner - Job Execution Engine
//!
//! Provides utilities for executing jobs (workflow steps) in a controlled manner.
//!
//! ## Responsibilities
//! - Execute jobs in defined order
//! - Control success/failure flow
//! - Handle process exit codes

use std::sync::Arc;

use tracing::{error, info, warn};

use crate::application::JobResult;
use crate::di::AppContext;
use crate::registry::{get_all_jobs, get_job};
use crate::shared::exit_codes;

/// Execute a single workflow step by name
pub async fn run_step(job_name: &str, ctx: Arc<AppContext>) -> Option<JobResult> {
    let job = match get_job(job_name) {
        Some(j) => j,
        None => {
            let available: Vec<_> = get_all_jobs().iter().map(|j| j.name.clone()).collect();
            error!(
                job_name = %job_name,
                available_jobs = ?available,
                "Workflow step not found"
            );
            return None;
        }
    };

    info!(job = %job_name, "Starting workflow step");
    Some(job.execute(ctx).await)
}

/// Execute all workflow steps in sequence
pub async fn run_workflow(ctx: Arc<AppContext>) -> i32 {
    let jobs = get_all_jobs();

    if jobs.is_empty() {
        warn!("No workflow steps registered");
        return exit_codes::SUCCESS;
    }

    let job_names: Vec<_> = jobs.iter().map(|j| j.name.clone()).collect();
    info!(steps = ?job_names, "Starting workflow");

    let mut has_failure = false;

    for job in jobs {
        let result = job.execute(ctx.clone()).await;
        if !result.success {
            has_failure = true;
            error!(job = %job.name, "Workflow step failed, continuing...");
        }
    }

    info!(success = %!has_failure, "Workflow completed");

    if has_failure {
        exit_codes::FAILURE
    } else {
        exit_codes::SUCCESS
    }
}

/// Execute a specific workflow step by name
pub async fn run_single_job(job_name: &str, ctx: Arc<AppContext>) -> i32 {
    match run_step(job_name, ctx).await {
        Some(result) => {
            if result.success {
                exit_codes::SUCCESS
            } else {
                exit_codes::FAILURE
            }
        }
        None => exit_codes::FAILURE,
    }
}
